src/core/gui/PagePreviewDecoration.cpp

```diff
@@ -0,0 +1,117 @@
+    switch (control->getSettings()->getSidebarNumberingStyle()) {
+        case SidebarNumberingStyle::STYLE_1:
+            PagePreviewDecoration::drawPageNumberBelowPreview(cr, pageEntry, control);
+            break;
+        case SidebarNumberingStyle::STYLE_2:
+            PagePreviewDecoration::drawPageNumberWithCircleBackground(cr, pageEntry, control);
+            break;
+        case SidebarNumberingStyle::STYLE_3:
+            PagePreviewDecoration::drawPageNumberWithSquareBackground(cr, pageEntry, control);
+            break;
+        default:
+            break;
+    }
+}
+
+void PagePreviewDecoration::drawPageNumberBelowPreview(cairo_t* cr, SidebarPreviewPageEntry* pageEntry,
+                                                       Control* control) {
+    xoj::util::CairoSaveGuard saveGuard(cr);
+    cairo_text_extents_t extents;
+    std::string pageNumber = std::to_string(pageEntry->getIndex() + 1);
+    Color color = control->getSettings()->getBackgroundColor();
+    if (pageEntry->isSelected()) {
+        color = control->getSettings()->getBorderColor();
+    }
+    Util::cairo_set_source_rgbi(cr, color);
+    cairo_select_font_face(cr, "Sans", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
+    cairo_set_font_size(cr, FONT_SIZE);
+    cairo_text_extents(cr, pageNumber.c_str(), &extents);
+    double x = pageEntry->getWidth() / 2 - (extents.width / 2 + extents.x_bearing);
+    double y = pageEntry->getHeight() - MARGIN_BOTTOM - (extents.height / 2 + extents.y_bearing);
+    cairo_move_to(cr, x, y);
+    cairo_show_text(cr, pageNumber.c_str());
+}
+
+void PagePreviewDecoration::drawPageNumberWithCircleBackground(cairo_t* cr, SidebarPreviewPageEntry* pageEntry,
+                                                               Control* control) {
+    xoj::util::CairoSaveGuard saveGuard(cr);
+    cairo_text_extents_t extents;
+    std::string pageNumber = std::to_string(pageEntry->getIndex() + 1);
+    cairo_select_font_face(cr, "Sans", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
+    cairo_set_font_size(cr, FONT_SIZE);
+    cairo_text_extents(cr, pageNumber.c_str(), &extents);
+
+    // Set the color for drawing the circle. This circle is required to provide good contrast for the page number.
+    if (pageEntry->isSelected()) {
+        Color color = control->getSettings()->getBorderColor();
+        Util::cairo_set_source_rgbi(cr, color, 0.6);
+    } else {
+        cairo_set_source_rgba(cr, 0.3, 0.3, 0.3, 0.6);
+    }
+
+    // Draw an slightly transparent circle with 11 pixel radius in the lower right corner of the preview
+    // The -33 is the indentation from the right border of the widget to the center of the circle
+    // The -36 is the indentation from the bottom border of the widget to the center of the circle
+    double x = pageEntry->getWidth() - 33, y = pageEntry->getHeight() - 36;
+    cairo_set_line_width(cr, 1.0);
+    cairo_arc(cr, x, y, 11, 0, 2 * M_PI);
```

Comment:
and all those constants 33, 36, 1.0 (the line width), 11 (the radius).

(0 and 2*M_PI are fine like that)